gnupg-pkcs11-scd/command.c

```diff
@@ -1134,127 +1201,169 @@ gpg_error_t _cmd_pksign_type (assuan_context_t ctx, char *line, int typehint)
 		error = GPG_ERR_INV_DATA;
 		goto cleanup;
 	}
-	/*
-	 * sender prefixed data with algorithm OID
-	 */
-	if (hash != NULL) {
-		if (!strcmp(hash, "rmd160") && data->size == (0x14 + sizeof(rmd160_prefix)) &&
-			!memcmp (data->data, rmd160_prefix, sizeof (rmd160_prefix))) {
-			inject = INJECT_NONE;
-		}
-		else if (!strcmp(hash, "rmd160") && data->size == 0x14) {
-			inject = INJECT_RMD160;
-		}
-		else if (!strcmp(hash, "md5") && data->size == (0x10 + sizeof(md5_prefix)) &&
-			!memcmp (data->data, md5_prefix, sizeof (md5_prefix))) {
-			inject = INJECT_NONE;
-		}
-		else if (!strcmp(hash, "md5") && data->size == 0x10) {
-			inject = INJECT_MD5;
-		}
-		else if (!strcmp(hash, "sha1") && data->size == (0x14 + sizeof(sha1_prefix)) &&
-			!memcmp (data->data, sha1_prefix, sizeof (sha1_prefix))) {
-			inject = INJECT_NONE;
-		}
-		else if (!strcmp(hash, "sha1") && data->size == 0x14) {
-			inject = INJECT_SHA1;
-		}
-		else if (!strcmp(hash, "sha224") && data->size == (0x1c + sizeof(sha224_prefix)) &&
-			!memcmp (data->data, sha224_prefix, sizeof (sha224_prefix))) {
-			inject = INJECT_NONE;
-		}
-		else if (!strcmp(hash, "sha224") && data->size == 0x1c) {
-			inject = INJECT_SHA224;
-		}
-		else if (!strcmp(hash, "sha256") && data->size == (0x20 + sizeof(sha256_prefix)) &&
-			!memcmp (data->data, sha256_prefix, sizeof (sha256_prefix))) {
-			inject = INJECT_NONE;
-		}
-		else if (!strcmp(hash, "sha256") && data->size == 0x20) {
-			inject = INJECT_SHA256;
-		}
-		else if (!strcmp(hash, "sha384") && data->size == (0x30 + sizeof(sha384_prefix)) &&
-			!memcmp (data->data, sha384_prefix, sizeof (sha384_prefix))) {
-			inject = INJECT_NONE;
-		}
-		else if (!strcmp(hash, "sha384") && data->size == 0x30) {
-			inject = INJECT_SHA384;
-		}
-		else if (!strcmp(hash, "sha512") && data->size == (0x40 + sizeof(sha512_prefix)) &&
-			!memcmp (data->data, sha512_prefix, sizeof (sha512_prefix))) {
-			inject = INJECT_NONE;
-		}
-		else if (!strcmp(hash, "sha512") && data->size == 0x40) {
-			inject = INJECT_SHA512;
-		}
-		else {
-			common_log (LOG_DEBUG, "unsupported hash algo (hash=%s,size=%d)", hash, data->size);
-			error = GPG_ERR_UNSUPPORTED_ALGORITHM;
+
+	if (
+		(error = _get_certificate_by_name (
+			ctx,
+			l,
+			typehint,
+			&cert_id,
+			NULL
+		)) != GPG_ERR_NO_ERROR
+	) {
+		goto cleanup;
+	}
+
+	if (
+		(error = get_cert_keyinfo(ctx, cert_id, &keyinfo)) != GPG_ERR_NO_ERROR
+	) {
+		goto cleanup;
+	}
+
+	if (_pkcs11_keyinfo_mechanism(keyinfo, &pkcs11_mechanism) != CKR_OK) {
+		goto cleanup;
```

Comment:
please focus in EC in this thread, any other change such as rename of the variables or modifying any behavior should go to other pull request.

In this case you assume that the problem is something within the token to select the hash, while as far as I remember the gnupg is the one that decides in some cases to pass the entire blob forcing us to guess what hash did he wish use to use.

anyway, in order to discuss it better please move this into other pull request.

Reply:
I'm not sure what you mean.

The highlighted change is regarding selecting the mechanism for asking the PKCS#11 module to sign using the correct mechanism, as well as ensuring PKCS padding is only used when using (`CKM_RSA_PKCS`)

Reply:
Maybe I do not understand the code, this is why it is better to do this in steps each change own pull request...

gnupg provides data which may or may not include the hash packaging, the fixup should be done in any case. 

you assume that this fixup should be done only if key type is rsa, correct? have you looked at the scd sources to see if this is correct and hash never provided in this case?

Reply:
doesn't ec also signs hashes? as far as I know it does. how will the peer know what hash was used in the signature? the hash name/oid must be signed as well for this to be secure.

Reply:
EcDSA keys for NIST P-256 are 33 bytes (257 bits) long (in compressed mode) with 64 byte signatures.

The `CKM_ECDSA` mechanism expects the input to already be hashed (https://www.cryptsoft.com/pkcs11doc/v220/group__SEC__12__3__6__ECDSA__WITHOUT__HASHING.html) and encoded to fit within 2*keyLength.

The padding/encoding for PKCS#1 isn't needed since the keys are so much smaller.

Reply:
if the digest algorithm id is not signed then I may modify the digest id in the unauthenticated message to less secure digest that its value matches the more secure so validate process will validate less secure digest without being aware of the switch. anyway, it is not that important as long as if you sign the gnupg peer will be able to verify.

Reply:
The code now removes any found prefix.